controllers/doguUpgradeManager.go

```diff
@@ -0,0 +1,159 @@
+	doguRemoteRegistry, err := cesremote.New(operatorConfig.GetRemoteConfiguration(), operatorConfig.GetRemoteCredentials())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new remote dogu registry: %w", err)
+	}
+
+	imageRegistry := imageregistry.NewCraneContainerImageRegistry(operatorConfig.DockerRegistry.Username, operatorConfig.DockerRegistry.Password)
+
+	restConfig := ctrl.GetConfigOrDie()
+	clientSet, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find cluster config: %w", err)
+	}
+	applier, _, err := apply.New(restConfig, k8sDoguOperatorFieldManagerName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create K8s applier: %w", err)
+	}
+	collectApplier := resource.NewCollectApplier(applier)
+
+	fileExtractor := newPodFileExtractor(client, restConfig, clientSet)
+
+	doguLocalRegistry := cesRegistry.DoguRegistry()
+
+	executor := resource.NewCommandExecutor(clientSet, clientSet.CoreV1().RESTClient())
+	serviceAccountCreator := serviceaccount.NewCreator(cesRegistry, executor)
+
+	df := cesregistry.NewLocalDoguFetcher(doguLocalRegistry)
+	rdf := cesregistry.NewResourceDoguFetcher(client, doguRemoteRegistry)
+
+	depValidator := dependency.NewCompositeDependencyValidator(operatorConfig.Version, doguLocalRegistry)
+	doguChecker := health.NewDoguChecker(client, df)
+	premisesChecker := upgrade.NewPremisesChecker(depValidator, doguChecker, doguChecker)
+
+	upgradeExecutor := upgrade.NewUpgradeExecutor(client, imageRegistry, collectApplier, fileExtractor, serviceAccountCreator, cesRegistry)
+
+	return &doguUpgradeManager{
+		client:              client,
+		eventRecorder:       eventRecorder,
+		localDoguFetcher:    df,
+		resourceDoguFetcher: rdf,
+		premisesChecker:     premisesChecker,
+		upgradeExecutor:     upgradeExecutor,
+	}, nil
+}
+
+type doguUpgradeManager struct {
+	// general purpose
+	client        client.Client
+	eventRecorder record.EventRecorder
+	// upgrade business
+	premisesChecker     premisesChecker
+	localDoguFetcher    localDoguFetcher
+	resourceDoguFetcher resourceDoguFetcher
+	upgradeExecutor     upgradeExecutor
+}
+
+func (dum *doguUpgradeManager) Upgrade(ctx context.Context, doguResource *k8sv1.Dogu) error {
+	err := doguResource.ChangeState(ctx, dum.client, k8sv1.DoguStatusUpgrading)
+	if err != nil {
+		return err
+	}
+
+	upgradeDoguName := doguResource.Spec.Name
+	upgradeDoguVersion := doguResource.Spec.Version
+
+	fromDogu, toDogu, developmentDoguMap, err := dum.getDogusForUpgrade(ctx, doguResource)
+	if err != nil {
+		return err
+	}
+
+	dum.normalEvent(doguResource, "Checking premises...")
+	err = dum.premisesChecker.Check(ctx, doguResource, fromDogu, toDogu)
+	if err != nil {
+		dum.errorEventf(doguResource, ErrorOnFailedPremisesUpgradeEventReason, "Checking premises failed: %s", err)
+		return fmt.Errorf("dogu upgrade %s:%s failed a premise check: %w", upgradeDoguName, upgradeDoguVersion, err)
```

Comment:
This event is redundant as it produces the following event:

```
Event ErrorOnUpgradePremise k8s-dogu-operator "Checking premises failed: ..."
```

The following error is always produced when an error is returned from the upgrade manager:

```
Event ErrorOnUpgrade k8s-dogu-operator "Dogu upgrade failed: dogu upgrade ldap:4.12.1-1 failed a premise check: ..."
```
Both events report the same failure. For this case we can remove the error event in the dogu upgrade manager.

 ```suggestion
		return fmt.Errorf("dogu upgrade %s:%s failed a premise check: %w", upgradeDoguName, upgradeDoguVersion, err)
```

Reply:
- [ ] Done